### PR TITLE
Add simple GCN model with Tox21 tutorials

### PR DESCRIPTION
**Pull request by nissy-dev**
## What I  did
- [x] implement GCN model
- [x] make Tox21 example
- [ ] refactor using Haiku -> i will make another PR

## Concern
- it is a little hard to manage model parameters by using only JAX
  - I consider about using [Haiku](https://github.com/deepmind/dm-haiku)
  - Haiku is a light wrapper for JAX and provide useful utilities for JAX modeling.

### CODE REVIEW

File: jaxchem/models/gcn.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    activation : activation function
+        Default to be relu function.
+    bias : bool
+        Whether to add bias after affine transformation, default to be True.
+    sparse : bool
+        Whether to use the matrix multiplication method for sparse matrix, default to be False.
+    W_init : initialize function for weight
+        Default to be He normal distribution.
+    b_init : initialize function for bias
+        Default to be normal distribution.
```
Comment (rbharath) [inline jaxchem/models/gcn.py:29]: Can you add a returns section to the docstring? This is formatted as:

```
Returns
--------
# Add description of returned quantities here.
```

File: jaxchem/models/gcn.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        the i-th GCN layer. ``len(batchnorm)`` equals the number of GCN layers. By default,
+        batch normalization is applied for all GCN layers.
+    dropout : list[float] or None
+        ``dropout[i]`` decides the dropout probability on the output of the i-th GCN layer.
+        ``len(dropout)`` equals the number of GCN layers. By default, no dropout is
+        performed for all layers.
+    bias : bool
+        Whether to add bias after affine transformation, default to be True.
+    sparse : bool
+        Whether to use the matrix multiplication method for sparse matrix,  default to be False.
```
Comment (rbharath) [inline jaxchem/models/gcn.py:78]: Same here

File: jaxchem/models/gcn_predicator.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    classifier_hidden_feats : int
+        Size of hidden graph representations in the classifier. Default to 128.
+    classifier_dropout : float
+        The probability for dropout in the classifier. Default to 0.
+    n_tasks : int
+        Number of tasks, which is also the output size. Default to 1.
+    bias : bool
+        Whether to add bias after affine transformation, default to be True.
+    sparse : bool
+        Whether to use the matrix multiplication method for sparse matrix,  default to be False.
```
Comment (rbharath) [inline jaxchem/models/gcn_predicator.py:38]: Same here

**Comment by vsomnath** [comment]
This looks good!

I would recommend you add tests to go along with this, just to see if the `init_fn` and `apply_fn` do what you want them to do. You can also include an overfitting test for the GCN model. 

I had a different question for @rbharath: As this a fresh repo, would it be better to include typed annotations?

**Comment by nd-02110114** [comment]
@vsomnath
Thanks! I will plan to write test codes in the next task 👍 

I finish implementing GCN model with Tox21 and I confirmed this model works well for only CPU.
And I rethink the implementation plan, so please check it in #1. 


Training log is here
```
/Users/nishikawadaiki/.pyenv/versions/miniconda3-latest/envs/deepchem/lib/python3.6/site-packages/jax/lib/xla_bridge.py:116: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
Starting training...
Iter 0/50 (7.5211 s) valid loss: 0.1952             valid roc_auc score: 0.2888
Iter 1/50 (4.9013 s) valid loss: 0.1910             valid roc_auc score: 0.3868
Iter 2/50 (4.8272 s) valid loss: 0.1621             valid roc_auc score: 0.4335
Iter 3/50 (5.3543 s) valid loss: 0.1685             valid roc_auc score: 0.5044
Iter 4/50 (4.8036 s) valid loss: 0.1527             valid roc_auc score: 0.4775
Iter 5/50 (4.8120 s) valid loss: 0.1669             valid roc_auc score: 0.3776
Iter 6/50 (4.8491 s) valid loss: 0.1780             valid roc_auc score: 0.5883
Iter 7/50 (8.3131 s) valid loss: 0.1651             valid roc_auc score: 0.5749
Iter 8/50 (7.5634 s) valid loss: 0.1774             valid roc_auc score: 0.6632
Iter 9/50 (10.0391 s) valid loss: 0.1519             valid roc_auc score: 0.6072
Iter 10/50 (5.4071 s) valid loss: 0.1539             valid roc_auc score: 0.6412
Iter 11/50 (9.5906 s) valid loss: 0.1520             valid roc_auc score: 0.6219
Iter 12/50 (6.2533 s) valid loss: 0.1649             valid roc_auc score: 0.4402
Iter 13/50 (9.4323 s) valid loss: 0.1520             valid roc_auc score: 0.6766
Iter 14/50 (9.6443 s) valid loss: 0.1517             valid roc_auc score: 0.6652
Iter 15/50 (7.9252 s) valid loss: 0.1518             valid roc_auc score: 0.4010
Iter 16/50 (4.8110 s) valid loss: 0.1646             valid roc_auc score: 0.5537
Iter 17/50 (4.9156 s) valid loss: 0.1516             valid roc_auc score: 0.6792
Iter 18/50 (5.3514 s) valid loss: 0.1516             valid roc_auc score: 0.5888
Iter 19/50 (4.8552 s) valid loss: 0.1519             valid roc_auc score: 0.6772
Iter 20/50 (5.0117 s) valid loss: 0.1517             valid roc_auc score: 0.5910
Iter 21/50 (4.8955 s) valid loss: 0.1642             valid roc_auc score: 0.6024
Iter 22/50 (4.9206 s) valid loss: 0.1514             valid roc_auc score: 0.6125
Iter 23/50 (6.6702 s) valid loss: 0.1515             valid roc_auc score: 0.5888
Iter 24/50 (9.6432 s) valid loss: 0.1642             valid roc_auc score: 0.6624
Iter 25/50 (5.4489 s) valid loss: 0.1661             valid roc_auc score: 0.4310
Iter 26/50 (5.0140 s) valid loss: 0.1642             valid roc_auc score: 0.6660
Iter 27/50 (5.7626 s) valid loss: 0.1645             valid roc_auc score: 0.6495
Early stopping...
Test roc_auc score: 0.7024
```

**Comment by rbharath** [comment]
@vsomnath That's a really good suggestion! I think we should start to move DeepChem broadly towards typed annotations if we can.

@nd-02110114 Could you look into https://github.com/python/mypy style annotations for your code? This would help make the code easier to maintain.

**Comment by nd-02110114** [comment]
@rbharath 
Yes! I will look into mypy. And, I added type annotation in a new plan.  
I seem that type annotation may work well because Haiku uses type annotation.

**Pull request merged**